src/JNI/com/amazonaws/kinesis/video/producer/jni/Parameters.cpp

```diff
@@ -1188,3 +1193,131 @@ BOOL releaseStreamDataBuffer(JNIEnv* env, jobject dataBuffer, UINT32 offset, PBY
+            CHK(javaStringLength <= MAX_IMAGE_PREFIX_LENGTH, STATUS_INVALID_ARG_LEN);
+            pStreamEventMetadata->imagePrefix = (PCHAR)MEMCALLOC(javaStringLength + 1, SIZEOF(CHAR));
+            retChars = env->GetStringUTFChars(retString, NULL);
+            STRNCPY(pStreamEventMetadata->imagePrefix, retChars, javaStringLength);
+            env->ReleaseStringUTFChars(retString, retChars);
+        }
+    }
+
+    methodId = env->GetMethodID(cls, "getNumberOfPairs", "()B");
+    if (methodId == NULL) {
+        DLOGW("Couldn't find method id getNumberOfPairs");
+    } else {
+        pStreamEventMetadata->numberOfPairs = env->CallByteMethod(streamEventMetadata, methodId);
+        CHK_JVM_EXCEPTION(env);
+    }
+
+    methodId = env->GetMethodID(cls, "getNames", "()[Ljava/lang/String;");
+    if (methodId == NULL) {
+        DLOGW("Couldn't find method id getNames");
+    } else {
+        if(!allocStreamEventMetadataArray(env, streamEventMetadata, pStreamEventMetadata->names, methodId)) {
+            DLOGW("Failed in allocStreamEventMetadataArray step for metadata names.");
+        }
+    }
+
+    methodId = env->GetMethodID(cls, "getValues", "()[Ljava/lang/String;");
+    if (methodId == NULL) {
+        DLOGW("Couldn't find method id getValues");
+    } else {
+        if(!allocStreamEventMetadataArray(env, streamEventMetadata, pStreamEventMetadata->values, methodId)) {
+            DLOGW("Failed in allocStreamEventMetadataArray step for metadata vlaues.");
+        }
+    }
+
+CleanUp:
+    return STATUS_FAILED(retStatus) ? FALSE : TRUE;
+}
+
+BOOL allocStreamEventMetadataArray(JNIEnv* env, jobject streamEventMetadata, PCHAR* metaDataArray, jmethodID methodId) {
+    STATUS retStatus = STATUS_SUCCESS;
+    jobjectArray retArray = NULL;
+    jsize namesArrayLength = 0;
+    jstring stringElement;
+    const char *retChars;
+
+    CHK(methodId != NULL, STATUS_NULL_ARG);
+    retArray = (jobjectArray) env->CallObjectMethod(streamEventMetadata, methodId);
+    CHK_JVM_EXCEPTION(env);
+    CHK(retArray != NULL, STATUS_NULL_ARG);
+
+    namesArrayLength = env->GetArrayLength(retArray);
+
+    // Verify array returned from Java is not too long.
+    // NOTE: This check will already be done in PIC, but can do it here to safely alloc/delloc memory here in JNI.
+    CHK(namesArrayLength <= MAX_EVENT_CUSTOM_PAIRS, STATUS_MAX_FRAGMENT_METADATA_COUNT);
+
+    // Null all elements for safety.
+    MEMSET(metaDataArray, NULL, sizeof(PCHAR) * MAX_EVENT_CUSTOM_PAIRS);
+            
+    // Iterate through the char pointers, allocating memory for the Java string that will be copied to the char pointers.
+    for (jsize i = 0; i < namesArrayLength; i++) {
+        stringElement = (jstring) env->GetObjectArrayElement(retArray, i);
+        jsize javaStringLength = env->GetStringUTFLength(stringElement);
+        CHK(javaStringLength >= 0, STATUS_INVALID_ARG); // (jsize is signed, but will be used as an unsigned SIZE_T)
```

Comment:
The `putKinesisVideoEventMetadata` API should allow values to empty or have length=0 because that is one of the ways to unset / remove persistent metadata. I'm not sure about name being empty but you may want to check that as well

Reply:
Btw, this _event_ metadata doesn't have the functionality of persistency, but it looks like I have already set it to allow a 0 value, maybe I did this after you added this comment. Can you please verify that `javaStringLength >= 0` is good to go and is what you were referring to, thanks.